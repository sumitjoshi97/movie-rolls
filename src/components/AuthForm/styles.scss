.auth-form {
  width: 46rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background: #fff;
  padding: 2rem 0;
  border-radius: 8px;

  &__logo {
    display: flex;
    justify-content: center;
    width: 30rem;
    padding-bottom: 1.4rem;
    color: #ff0000;
    font-weight: 700;

    &__main {
      font-size: 4rem;
    }
    &__sub {
      font-size: 2rem;
    }
  }

  &__header {
    color: #333;
    font-weight: 700;
    width: 30rem;

    & > h3 {
      font-size: 3.3rem;
      text-align: center;
    }
  }

  &__form {
    display: flex;
    flex-direction: column;
    width: 30rem;

    justify-content: center;
    align-items: center;
    padding-top: 2rem;

    &__input {
      width: 100%;
      font-size: 1.6rem;
      padding: 1.4rem;
      margin-bottom: 1.4rem;
      border-radius: 4px;
      border: 1px solid #979797;
    }

    &__submit {
      background: #ff0000;
    }
  }

  & > span {
    color: #333;
    font-weight: 700;
    padding-bottom: 1.2rem;
  }

  &__social-btns {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 38rem;

    &__google {
      background: #4285f4;
      text-decoration: none;
      text-align: center;
    }

    &__facebook {
      background: #3b5998;
      margin-bottom: 3rem;
    }
  }

  &__active {
    background: none;
    border: none;
    color: #555555;
    font-weight: 700;
    padding-top: 1.4rem;
    padding-bottom: 0.4rem;
    cursor: pointer;

    &:hover {
      text-decoration: underline;
    }

    &:active,
    &:focus {
      outline: none;
    }
  }
}

.action-btn {
  font-size: 1.8rem;
  width: 30rem;

  padding: 1.4rem;
  color: #fff;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  margin-bottom: 1.4rem;
}
