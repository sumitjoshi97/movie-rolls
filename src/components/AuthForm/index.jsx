import React, { Component } from 'react'
import { connect } from 'react-redux'

import * as actions from '../../actions'

import './styles.scss'

class AuthForm extends Component {
  state = {
    email: '',
    password: '',
    active: 'login',
    redirect: false
  }

  handleInput = e => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handleActiveAuth = () => {
    const { active } = this.state
    if (active === 'login') {
      this.setState({
        active: 'signup'
      })
    } else {
      this.setState({
        active: 'login'
      })
    }
  }

  authWithEmailPassword = async e => {
    e.preventDefault()
    const { email, password, active } = this.state
    const { signupUser, loginUserWithEmail, fetchUser } = this.props

    if (active === 'login') {
      await loginUserWithEmail(email, password)
      await fetchUser()

      if (this.props.user) {
        await this.props.fetchList('favorite', this.props.user.uid)
        await this.props.fetchList('watch', this.props.user.uid)
      }
    } else {
      await signupUser(email, password)
      await fetchUser()
      if (this.props.user) {
        await this.props.fetchList('favorite', this.props.user.uid)
        await this.props.fetchList('watch', this.props.user.uid)
      }
    }
  }

  authWithSocial = async provider => {
    const { user } = this.props
    await this.props.loginUserWithSocial(provider)
    await this.props.fetchUser()
    if (user) {
      await this.props.fetchList('favorite', user.uid)
      await this.props.fetchList('watch', user.uid)
    }
  }

  render() {
    const { active } = this.state
    return (
      <div className='auth-form'>
        <div className='auth-form__logo'>
          <span className='auth-form__logo__main'>M</span>
          <span className='auth-form__logo__sub'>r</span>
        </div>
        <div className='auth-form__header'>
          <h3>{active === 'login' ? 'Log in' : 'Sign up'} to see more</h3>
        </div>

        <form className='auth-form__form'>
          <input
            type='email'
            name='email'
            className='auth-form__form__input'
            placeholder='Email address'
            value={this.state.email}
            onChange={this.handleInput}
          />

          <input
            type='password'
            name='password'
            className='auth-form__form__input'
            placeholder='Password'
            value={this.state.password}
            onChange={this.handleInput}
          />

          <input
            type='submit'
            className='auth-form__form__submit action-btn'
            value={active === 'login' ? 'Log in' : 'Sign up'}
            onClick={this.authWithEmailPassword}
          />
        </form>

        <span>Or</span>

        <div className='auth-form__social-btns'>
          <button
            className='auth-form__social-btns__google action-btn'
            onClick={() => this.authWithSocial('google')}
          >
            Continue with Google
          </button>
          <button
            className='auth-form__social-btns__facebook action-btn'
            onClick={() => this.authWithSocial('facebook')}
          >
            Continue with Facebook
          </button>
        </div>

        <button className='auth-form__active' onClick={this.handleActiveAuth}>
          {active === 'login'
            ? 'Not on Movie Rolls? Sign up'
            : 'Already a member? Login'}
        </button>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  user: state.auth.user
})

const mapDispatchToProps = dispatch => ({
  signupUser: (email, password) =>
    dispatch(actions.createUser(email, password)),

  loginUserWithEmail: (email, password) =>
    dispatch(actions.loginUserWithEmailPassword(email, password)),

  loginUserWithSocial: provider =>
    dispatch(actions.loginUserWithSocial(provider)),

  fetchUser: () => dispatch(actions.fetchUser()),
  fetchList: (type, userId) => dispatch(actions.fetchList(type, userId))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AuthForm)
