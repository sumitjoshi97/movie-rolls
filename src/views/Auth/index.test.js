import React from 'react'
import { shallow } from 'enzyme'
import { Redirect } from 'react-router-dom'
import { Auth } from './index'
import AuthForm from '../../components/AuthForm'

describe('Auth view unit tests', () => {
  describe('when isAuth is false', () => {
    const props = {
      isAuth: false,
      redirectPath: '/'
    }

    const authWrapper = shallow(<Auth {...props} />)

    test('should match snapshot', () => {
      expect(authWrapper).toMatchSnapshot()
    })

    test('should render `Redirect` when isAuth is false', () => {
      expect(authWrapper.find(Redirect)).toBeTruthy()
    })
  })

  describe('when isAuth is true', () => {
    const props = {
      isAuth: false,
      redirectPath: '/'
    }

    const authWrapper = shallow(<Auth {...props} />)

    test('should render `AuthForm` when isAuth is true', () => {
      expect(authWrapper.find(AuthForm)).toBeTruthy()
    })
  })
})
