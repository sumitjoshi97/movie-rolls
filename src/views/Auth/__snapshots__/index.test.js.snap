// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth view unit tests when isAuth is false should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Auth
    isAuth={false}
    redirectPath="/"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="auth__logo"
        >
          movie rolls
        </div>,
        null,
        <Connect(AuthForm) />,
      ],
      "className": "auth",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "movie rolls",
          "className": "auth__logo",
        },
        "ref": null,
        "rendered": "movie rolls",
        "type": "div",
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="auth__logo"
          >
            movie rolls
          </div>,
          null,
          <Connect(AuthForm) />,
        ],
        "className": "auth",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "movie rolls",
            "className": "auth__logo",
          },
          "ref": null,
          "rendered": "movie rolls",
          "type": "div",
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
